src/delay/timer.rs

```diff
@@ -0,0 +1,193 @@
+//! Delay implementation based on general-purpose 32 bit timers.
+//!
+//! TIM2 and TIM5 are a general purpose 32-bit auto-reload up/downcounter with
+//! a 16-bit prescaler.
+
+use core::cmp::max;
+
+use cast::{u16, u32};
+use embedded_hal::blocking::delay::{DelayMs, DelayUs};
+
+use crate::{
+    bb,
+    pac::{self, RCC},
+    rcc::Clocks,
+};
+
+macro_rules! hal {
+    ($($TIM:ident: ($struct:ident, $waitfn:ident, $en_bit:expr, $apbenr:ident, $apbrstr:ident, $pclk:ident, $ppre:ident),)+) => {
+        $(
+            /// General purpose timer as delay provider
+            pub struct $struct {
+                clocks: Clocks,
+                tim: pac::$TIM,
+            }
+
+            fn $waitfn(tim: &mut pac::$TIM, prescaler: u16, auto_reload_register: u32) {
+                // Write Prescaler (PSC)
+                tim.psc.write(|w| w.psc().bits(prescaler));
+
+                // Write Auto-Reload Register (ARR)
+                // Note: Make it impossible to set the ARR value to 0, since this
+                // would cause an infinite loop.
+                tim.arr.write(|w| unsafe { w.bits(max(1, auto_reload_register)) });
+
+                // Trigger update event (UEV) in the event generation register (EGR)
+                // in order to immediately apply the config
+                tim.cr1.modify(|_, w| w.urs().set_bit());
+                tim.egr.write(|w| w.ug().set_bit());
+                tim.cr1.modify(|_, w| w.urs().clear_bit());
+
+                // Configure the counter in one-pulse mode (counter stops counting at
+                // the next updateevent, clearing the CEN bit) and enable the counter.
+                tim.cr1.write(|w| w.opm().set_bit().cen().set_bit());
+
+                // Wait for CEN bit to clear
+                while tim.cr1.read().cen().is_enabled() { /* wait */ }
+            }
+
+            impl $struct {
+                /// Configures the timer as a delay provider
+                pub fn new(tim: pac::$TIM, clocks: Clocks) -> Self {
+                    unsafe {
+                        //NOTE(unsafe) this reference will only be used for atomic writes with no side effects
+                        let rcc = &(*RCC::ptr());
+
+                        // Enable timer peripheral in RCC
+                        bb::set(&rcc.$apbenr, $en_bit);
+
+                        // Stall the pipeline to work around erratum 2.1.13 (DM00037591)
+                        cortex_m::asm::dsb();
+
+                        // Reset timer
+                        bb::set(&rcc.$apbrstr, $en_bit);
+                        bb::clear(&rcc.$apbrstr, $en_bit);
+                    }
```

Comment:
These lines are re-used a few times within the codebase, not sure if it might make sense to create an internal helper function for them...